src/ndb_p1.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+
+#include <string.h>
+#include "ndb.h"
+
+ephemeris_t            ndb_ephemeris[PLATFORM_SIGNAL_COUNT];
+ndb_element_metadata_t ndb_ephemeris_md[PLATFORM_SIGNAL_COUNT];
+almanac_t              ndb_almanac[PLATFORM_SIGNAL_COUNT];
+ndb_element_metadata_t ndb_almanac_md[PLATFORM_SIGNAL_COUNT];
+
+#define NDB_EPHE_FILE_NAME   "ephemerides"
+#define NDB_ALMA_FILE_NAME   "almanacs"
+
+ndb_file_t ndb_ephe_file = {
+    .name = NDB_EPHE_FILE_NAME,
+    .fh = -1
+};
+ndb_file_t ndb_alma_file = {
+    .name = NDB_ALMA_FILE_NAME,
+    .fh = -1
+};
+
+enum ndb_op_code ndb_p1_init()
+{
+    if(ndb_load_data(&ndb_ephe_file, ndb_ephemeris, ndb_ephemeris_md,
+                      sizeof(ephemeris_t), PLATFORM_SIGNAL_COUNT)
+            != NDB_ERR_NONE) {
+        /* TODO: log failure */
+    }
+
+    if(ndb_load_data(&ndb_alma_file, ndb_almanac, ndb_almanac_md,
+                      sizeof(almanac_t), PLATFORM_SIGNAL_COUNT)
+            != NDB_ERR_NONE) {
+        /* TODO: log failure */
+    }
+    return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_ephemeris_read(gnss_signal_t *sid, ephemeris_t *e)
+{
+    u16 idx = sid_to_global_index(*sid);
+    if(!ndb_ephemeris[idx].valid) {
+        return NDB_ERR_MISSING_IE;
+    }
+    ndb_retrieve(e, &ndb_ephemeris[idx], sizeof(ephemeris_t));
+    return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_ephemeris_store(gnss_signal_t *sid, ephemeris_t *e)
+{
+    u16 idx = sid_to_global_index(*sid);
+    return ndb_update(&ndb_ephemeris[idx], e, sizeof(ephemeris_t),
+                      &ndb_ephemeris_md[idx]);
+}
+
+enum ndb_op_code ndb_almanac_read(gnss_signal_t *sid, almanac_t *a)
+{
+    u16 idx = sid_to_global_index(*sid);
+    if(!ndb_almanac[idx].valid) {
+        return NDB_ERR_MISSING_IE;
+    }
+    ndb_retrieve(a, &ndb_almanac[idx], sizeof(almanac_t));
+    return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_almanac_store(gnss_signal_t *sid, almanac_t *a)
+{
+    u16 idx = sid_to_global_index(*sid);
+    return ndb_update(&ndb_almanac[idx], a, sizeof(almanac_t),
+                      &ndb_almanac_md[idx]);
+}
+
+void ndb_p1_nvm_write()
+{
+    int idx;
+    for(idx = 0; idx < PLATFORM_SIGNAL_COUNT; idx++) {
+        if(ndb_ephemeris[idx].valid && !ndb_ephemeris_md[idx].saved){
+            ephemeris_t t_ephe;
+            ndb_lock(1);
```

Comment:
Or... maybe some way to expose the data elements to the core when you register a section so that it can autonomously scan through them and write to NVM as needed.